### PR TITLE
AAE-8817 filter users by application keycloack

### PR DESCRIPTION
**Pull request by Giovanni007**
Keyclaok implementation to filter users by application
Ref: https://github.com/Activiti/Activiti/issues/3943

### CODE REVIEW

File: activiti-cloud-service-common/activiti-cloud-services-common-identity-keycloak/src/main/java/org/activiti/cloud/services/identity/keycloak/KeycloakManagementService.java [hunk 1]
```diff
@@ -42,28 +47,61 @@ public class KeycloakManagementService implements IdentityManagementService {
...
+            .collect(Collectors.toList());
+
+        if(!StringUtils.isEmpty(userSearchParams.getApplication())) {
+            return filterUsersInApplicationsScope(users, userSearchParams);
+        } else {
+            return filterUsers(users, userSearchParams);
+        }
+    }
+
+    private List<User> filterUsers(List<User> users ,UserSearchParams userSearchParams) {
```
Comment (miguelruizdev) [inline activiti-cloud-service-common/activiti-cloud-services-common-identity-keycloak/src/main/java/org/activiti/cloud/services/identity/keycloak/KeycloakManagementService.java:79]: just to double-check, the idea behind the `filterUsers` and `filterUsersInApplicationsScope` is that we provide both filtering options for single & multiple application?
in the same way we offer the possibility of setting `keycloak.use-resource-role-mappings` to both `false` and `true` ?

Comment (Giovanni007) [reply activiti-cloud-service-common/activiti-cloud-services-common-identity-keycloak/src/main/java/org/activiti/cloud/services/identity/keycloak/KeycloakManagementService.java:79]: If with "single application" you mean the `realm` level roles, yes.
In the basic scenario, we search for users and if we want to filter by their role we check the realm roles.
Instead, If we pass the parameter `application` then we return:

- the users that have access to the application (have at least one role in the Keycloack "client" role mapping)
- if we pass application and roles, then we return users that have that role in the client ignoring the REALM roles

Comment (miguelruizdev) [reply activiti-cloud-service-common/activiti-cloud-services-common-identity-keycloak/src/main/java/org/activiti/cloud/services/identity/keycloak/KeycloakManagementService.java:79]: Understood, thanks for the explanation 👌

**Review by miguelruizdev — APPROVED** [review]

File: activiti-cloud-service-common/activiti-cloud-services-common-identity-keycloak/src/main/java/org/activiti/cloud/services/identity/keycloak/KeycloakManagementService.java (hunk 1 above)
Comment (mauriziovitale) [reply activiti-cloud-service-common/activiti-cloud-services-common-identity-keycloak/src/main/java/org/activiti/cloud/services/identity/keycloak/KeycloakManagementService.java:79]: Are we taking into account multiple apps ?

Comment (Giovanni007) [reply activiti-cloud-service-common/activiti-cloud-services-common-identity-keycloak/src/main/java/org/activiti/cloud/services/identity/keycloak/KeycloakManagementService.java:79]: No, the parameter `application` is a String, not a list.

**Review by balsarori — APPROVED** [review]
LGTM

File: activiti-cloud-service-common/activiti-cloud-services-common-identity-keycloak/src/main/java/org/activiti/cloud/services/identity/keycloak/KeycloakManagementService.java [hunk 2]
```diff
@@ -42,28 +47,61 @@ public class KeycloakManagementService implements IdentityManagementService {
...
             .collect(Collectors.toList());
     }
 
+    private List<User> filterUsersInApplicationsScope(List<User> users ,UserSearchParams userSearchParams) {
+        String application = userSearchParams.getApplication();
+        String kClientId = getKeycloakClientId(application);
+        if(StringUtils.isEmpty(kClientId)) {
+            return Collections.emptyList();
+        }
+       users.forEach(user -> user.setApplicationRoles(
```
Comment (mauriziovitale) [inline activiti-cloud-service-common/activiti-cloud-services-common-identity-keycloak/src/main/java/org/activiti/cloud/services/identity/keycloak/KeycloakManagementService.java:93]: do we really need it?

**Review by mauriziovitale — CHANGES_REQUESTED** [review]
check the comment

File: activiti-cloud-service-common/activiti-cloud-services-common-identity-keycloak/src/main/java/org/activiti/cloud/services/identity/keycloak/KeycloakManagementService.java (hunk 2 above)
Comment (Giovanni007) [reply activiti-cloud-service-common/activiti-cloud-services-common-identity-keycloak/src/main/java/org/activiti/cloud/services/identity/keycloak/KeycloakManagementService.java:93]: I'm going to remove it in another PR

**Review by mauriziovitale — APPROVED** [review]

**Pull request merged**